File: apps/backend/scripts/patch-n8n-stage1-phase15.mjs
/**
 * Patches live Campaign Compare Stage 1 workflow (tk2tAop5hzSjGSqv) for Phase 1.5:
 * - LLM system prompt (rich report schema)
 * - Build Callback Body (passes Phase 1.5 fields)
 * - Callback HTTP jsonBody
 * - Format Email (rich text sections)
 *
 * Run on VPS:
 *   node patch-n8n-stage1-phase15.mjs /root/n8n-data-old/database.sqlite
 */
import Database from 'better-sqlite3';
import {
    BUILD_CALLBACK_CODE,
    CALLBACK_JSON_BODY,
    FORMAT_EMAIL_CODE,
    LLM_SYSTEM,
} from './campaign-compare-stage1-phase15-prompt.mjs';

const dbPath = process.argv[2] || '/root/n8n-data-old/database.sqlite';
const WF_ID = 'tk2tAop5hzSjGSqv';

const db = new Database(dbPath);
const row = db.prepare('SELECT nodes, connections, activeVersionId FROM workflow_entity WHERE id = ?').get(WF_ID);
if (!row) throw new Error(`workflow ${WF_ID} not found`);

const nodes = JSON.parse(row.nodes);

const llm = nodes.find((n) => n.name === 'Compare Stage 1 LLM');
const build = nodes.find((n) => n.name === 'Build Callback Body');
const callback = nodes.find((n) => n.name === 'Callback to Backend');
const formatEmail = nodes.find((n) => n.name === 'Format Email');

if (!llm || !build || !callback) {
    throw new Error('missing Compare Stage 1 LLM / Build Callback Body / Callback to Backend nodes');
}

if (llm.parameters?.messages?.messageValues?.[0]) {
    llm.parameters.messages.messageValues[0].message = LLM_SYSTEM;
}

build.parameters.jsCode = BUILD_CALLBACK_CODE;
callback.parameters.jsonBody = CALLBACK_JSON_BODY;

if (formatEmail) {
    formatEmail.parameters.jsCode = FORMAT_EMAIL_CODE;
}

const nodesJson = JSON.stringify(nodes);
const connectionsJson = row.connections;

db.prepare('UPDATE workflow_entity SET nodes = ? WHERE id = ?').run(nodesJson, WF_ID);

const activeVersionId = row.activeVersionId;
if (activeVersionId) {
    db.prepare('UPDATE workflow_history SET nodes = ? WHERE versionId = ?').run(nodesJson, activeVersionId);
}

console.log('patched Phase 1.5 prompt on workflow', WF_ID, 'activeVersionId', activeVersionId);
console.log('updated nodes:', ['Compare Stage 1 LLM', 'Build Callback Body', 'Callback to Backend', formatEmail ? 'Format Email' : '(no Format Email)'].join(', '));
db.close();

File: apps/backend/scripts/patch-n8n-stage1-phase15.py
#!/usr/bin/env python3
"""Patch live n8n Stage 1 workflow for Phase 1.5 (stdlib sqlite3 only)."""
import json
import re
import sqlite3
import sys
from pathlib import Path

DB = sys.argv[1] if len(sys.argv) > 1 else '/root/n8n-data-old/database.sqlite'
WF_ID = 'tk2tAop5hzSjGSqv'
MJS = Path(__file__).resolve().parent / 'campaign-compare-stage1-phase15-prompt.mjs'


def extract_backtick(name: str, text: str) -> str:
    m = re.search(rf"export const {name} = `(.*?)`;", text, re.S)
    if not m:
        raise ValueError(f'missing export {name}')
    return m.group(1)


def extract_single_quoted(name: str, text: str) -> str:
    m = re.search(rf"export const {name} =\s*\n?\s*'(.*?)';", text, re.S)
    if not m:
        raise ValueError(f'missing export {name}')
    return m.group(1)


def main() -> None:
    text = MJS.read_text(encoding='utf-8')
    llm_system = extract_backtick('LLM_SYSTEM', text)
    build_code = extract_backtick('BUILD_CALLBACK_CODE', text)
    format_email = extract_backtick('FORMAT_EMAIL_CODE', text)
    callback_json = extract_single_quoted('CALLBACK_JSON_BODY', text)

    conn = sqlite3.connect(DB)
    row = conn.execute(
        'SELECT nodes, activeVersionId FROM workflow_entity WHERE id = ?', (WF_ID,)
    ).fetchone()
    if not row:
        raise SystemExit(f'workflow {WF_ID} not found')

    nodes = json.loads(row[0])
    llm = next(n for n in nodes if n.get('name') == 'Compare Stage 1 LLM')
    build = next(n for n in nodes if n.get('name') == 'Build Callback Body')
    callback = next(n for n in nodes if n.get('name') == 'Callback to Backend')
    format_node = next((n for n in nodes if n.get('name') == 'Format Email'), None)

    llm['parameters']['messages']['messageValues'][0]['message'] = llm_system
    build['parameters']['jsCode'] = build_code
    callback['parameters']['jsonBody'] = callback_json
    if format_node:
        format_node['parameters']['jsCode'] = format_email

    nodes_json = json.dumps(nodes, ensure_ascii=False)
    conn.execute('UPDATE workflow_entity SET nodes = ? WHERE id = ?', (nodes_json, WF_ID))
    active_version = row[1]
    if active_version:
        conn.execute(
            'UPDATE workflow_history SET nodes = ? WHERE versionId = ?',
            (nodes_json, active_version),
        )
    conn.commit()
    conn.close()
    print('patched Phase 1.5 on', WF_ID, 'activeVersionId', active_version)


if __name__ == '__main__':
    main()
